### ISSUE TITLE
Support battery storage

### ISSUE THREAD

**Issue by davet2001** [davet2001/energy-sankey#16]

**Comment by 18107** [comment]
Batteries do not display properly.
![Energy-sankey batteries](https://github.com/user-attachments/assets/c15d2476-3293-4dd0-9c0a-b2a29310dba0)

**Comment by davet2001** [comment]
@18107 Thanks for sharing your screenshot!

Yes, you are correct, batteries are not supported yet.

That is a feature I'm working on but will take some time. My plan is for this to be at the bottom of the diagram in the middle.

Since I don't have batteries to test, please could you help me by clarifying how they are represented in HA?

Do they operate as a single power entity that can consume a positive amount or produce a negative amount? Or are the in/out handled as two separate entities?

Thanks in advance for your help!

**Comment by davet2001** [comment]
Ok so do you have 2 batteries which can each operate independently?

Need to think of a way to diagrammatically represent that... so far is was just thinking of 1.

**Comment by 18107** [comment]
I'd suggest combining them, but either battery can charge the other. 

Moving them wouldn't be too bad. Left side when discharging and right side when charging.

**Comment by davet2001** [comment]
Ok, the bigger issue is more of a visual one. If there is only one battery, it can go at the bottom, with no lines crossing. If there are more than one, the sankey lines could cross on an energy graph, as we'd need to show an input and output energy for each simultaneously, and presumably recombine them.

Power graphs are relatively easy because the flow is only going in one direction for every item at a time. On an energy graph, in a period there can be both positive and negative flow. Most common example is solar panels, which might cause the grid interface to be exporting during the day and importing during the night. This is currently handled well for solar panels, but not for batteries yet.

I've got a couple of ideas, and will try to sketch them later.

**Comment by davet2001** [comment]
Ok, here's a first pass at something possible. It's hand drawn with overlays to represent the colours. 

The numbers are all made up.
![image](https://github.com/user-attachments/assets/6ba0d488-6fa8-4427-8648-300b93889fa7)

First impression is that it looks complicated.

But if you look again, is there a better way to present this? 

The image shown is probably the worst case, where an energy diagram can show in and out of the grid & batteries, with all sources flowing to all destinations.

Feedback most welcome. I'll attach the inkscape file in case you want to annotate it.

**Comment by 18107** [comment]
I think I'd prioritise solar going to load, then any to the grid would be from batteries first. That would better show that exports could be reduced by reducing battery output.

The graph doesn't show power moving from one battery to the other. In my poorly optimised system this is actually quite common, despite all of my attempts to stop it. There are still a few rare situations where I want to directly transfer power between batteries.

My only other idea would be to separate the input and output for each device into it's own entity on the graph. That way all power can flow from left to right. The downside is that it becomes 2 dimensional.
I do like your solution more for an energy overview. It's a much better graph than I made.

**Comment by BJReplay** [comment]
> Do they operate as a single power entity that can consume a positive amount or produce a negative amount? Or are the in/out handled as two separate entities?

Hi @davet2001 - love your integration.

In my case, I have a battery (Powerwall 2, with Gateway 1) that represents the power as two entities (Import - to the battery, and Export - from the battery).

Similarly the grid has two entities.

**Comment by davet2001** [comment]
@BJReplay Noted, thanks!
Are there two entities for power as well as two entities for energy?

**Comment by 18107** [comment]
I have 2 entities for energy (kWh) and 1 entity for power (kW).

**Comment by BJReplay** [comment]
> Are there two entities for power as well as two entities for energy?

Yes, but on looking further, there is a combined  (bidirectional) entity for Power as well as the export and import separated entities.

So, at the simplest level - one power, two energy for each of grid, battery, and one each (for power and energy) for home load (which can be derived - and is, I believe internally) and one for solar.

**Comment by BJReplay** [comment]
> Ok, it's taken quite a bit of work to get to this point, but nearly there now.
> Sneak peek:

Happy to beta (or alpha) test.  Let me know if there's another repo or branch that has code you want to test.  As an active developer and owner of an HA integration, I'm familiar with testing unreleased code.

**Comment by davet2001** [comment]
@BJReplay Version 0.0.14 is now available as pre-release from HACS, supporting batteries for both power and energy graphs.

Please try it out and let me know how it goes!
